internal/sdk/crypto/crypto.go

```diff
@@ -0,0 +1,222 @@
+package crypto
+
+import (
+	"bytes"
+	"context"
+	"crypto/md5"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	sdkHttp "github.com/altinity/terraform-provider-altinitycloud/internal/sdk/http"
+)
+
+type Crypto struct {
+	RootCAs *x509.CertPool
+	URL     string
+}
+
+func NewCrypto(rootCAs *x509.CertPool, cryptoUrl string) *Crypto {
+	return &Crypto{
+		RootCAs: rootCAs,
+		URL:     cryptoUrl,
+	}
+}
+
+var (
+	pemEndPrefix = []byte("-----END ")
+	pemEndSuffix = []byte("-----")
+)
+
+func (c *Crypto) Encrypt(pem string, value string) (string, error) {
+	split := split([]byte(pem))
+	if len(split) != 2 {
+		return "", fmt.Errorf("malformed %s: expected 2 PEMs, instead got %d", pem, len(split))
+	}
+	tlsCert, err := x509KeyPairWithLeaf(split[0], split[1])
+	if err != nil {
+		return "", err
+	}
+
+	res, err := c.fetchPublicKey(context.Background(), tlsCert)
+	if err != nil {
+		return string(split[0]) + string(split[1]), err
+	}
+	key, err := parseRSAPublicKey(res)
+	if err != nil {
+		return "", err
+	}
+	v, err := encryptWithRSAPublicKey(value, key)
+	if err != nil {
+		return "", err
+	}
+
+	return v, nil
+}
+
+func (c *Crypto) fetchPublicKey(ctx context.Context, tlsCert tls.Certificate) (pem []byte, err error) {
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: c.RootCAs,
+				Certificates: []tls.Certificate{
+					tlsCert,
+				},
+			},
+		},
+		Timeout: time.Minute,
+	}
+	url := fmt.Sprintf("%s/key", c.URL)
```

Comment:
Consider pinning certs or having some other mechanism to prevent injection of an attacker-controlled key.

Reply:
I am not sure how can I pin certificates here. Certs are dynamically generated by control-plane as the user request them. What am I missing?